library/kernel_settings.py

```diff
@@ -0,0 +1,853 @@
+
+TUNED_PROFILE = os.environ.get("TEST_PROFILE", "kernel_settings")
+NOCHANGES = 0
+CHANGES = 1
+REMOVE_SECTION_VALUE = {"state": "empty"}
+SECTION_TO_REPLACE = "__section_to_replace"
+ERR_SECTION_MISSING_NAME = "Error: section [{}] item is missing 'name': {}"
+ERR_NAME_NOT_VALID = "Error: section [{}] item name [{}] is not a valid string"
+ERR_NO_VALUE_OR_STATE = (
+    "Error: section [{}] item name [{}] must have either a " "'value' or 'state'"
+)
+ERR_BOTH_VALUE_AND_STATE = (
+    "Error: section [{}] item name [{}] must have only one of " "'value' or 'state'"
+)
+ERR_STATE_ABSENT = (
+    "Error: section [{}] item name [{}] state value must be " "'absent' not [{}]"
+)
+ERR_UNEXPECTED_VALUES = "Error: section [{}] item [{}] has unexpected values {}"
+ERR_VALUE_ITEM_NOT_DICT = "Error: section [{}] item name [{}] value [{}] is not a dict"
+ERR_VALUE_ITEM_PREVIOUS = (
+    "Error: section [{}] item name [{}] has invalid value for 'previous' [{}]"
+)
+ERR_REMOVE_SECTION_VALUE = "Error: to remove the section [{}] specify the value {}"
+ERR_ITEM_NOT_DICT = "Error: section [{}] item value [{}] is not a dict"
+ERR_ITEM_PREVIOUS = "Error: section [{}] item has invalid value for 'previous' [{}]"
+ERR_ITEM_DICT_OR_LIST = "Error: section [{}] value must be a dict or a list"
+ERR_LIST_NOT_ALLOWED = "Error: section [{}] item [{}] has unexpected list value {}"
+ERR_BLCMD_MUST_BE_LIST = "Error: section [{}] item [{}] must be a list not [{}]"
+
+
+def get_tuned_plugin_names():
+    """get names of all supported tuned plugins"""
+    # this relies on some assumptions:
+    # - the __init__ for all of these classes really don't do anything
+    # - load_all_plugins doesn't need any of the factory/providers from the __init__
+    # - the plugin_class __init__ really doesn't do anything
+    # - the name property doesn't use/do anything except return a name string
+    if not getattr(tuned.plugins.repository.Repository, "load_all_plugins", False):
+        # method added in tuned 2.11 - before that, just return hardcoded list
+        return [
+            "audio",
+            "bootloader",
+            "cpu",
+            "disk",
+            "eeepc_she",
+            "modules",
+            "mounts",
+            "net",
+            "script",
+            "scsi_host",
+            "selinux",
+            "sysctl",
+            "sysfs",
+            "systemd",
+            "usb",
+            "video",
+            "vm",
+        ]
+    names = []
+    storage_provider = tuned.storage.PickleProvider()
+    storage_factory = tuned.storage.Factory(storage_provider)
+    plugins_repo = tuned.plugins.repository.Repository(
+        None, None, None, None, None, None, None, None
+    )
+    for plugin_class in plugins_repo.load_all_plugins():
+        try:
+            tmpplg = plugin_class(
+                None, storage_factory, None, None, None, None, None, None
+            )
+            names.append(tmpplg.name)
+        except NotSupportedPluginException:
```

Comment:
The question is, do you want to list unsupported plugins?

Reply:
Yes, I think so, if possible.  For example - the user provides a single set of values for multiple platforms, and the kernel_settings/tuned module applies only those plugins that are supported on the specific platform, and ignores the other plugins.

If this isn't possible, then the user will have to set platform specific kernel settings - also doable, just not as elegant.

Reply:
Sounds OK to me. I just wanted to ask.